DOCS-1335: Add queryable sensor data docs (#2197)
viamrobotics/docs: docs/services/data/configure-data-query.md

1. If you haven't already, [download the `mongosh` shell](https://www.mongodb.com/try/download/shell).
   See the [`mongosh` documentation](https://www.mongodb.com/docs/mongodb-shell/) for more information.

1. Run the following command to connect to the Viam organization's MongoDB Atlas instance from `mongosh`:

   ```sh {class="line-numbers linkable-line-numbers"}
   mongosh "mongodb+srv://<YOUR-DB-HOSTNAME>" --apiVersion 1 --username <YOUR-DB-USER>
   ```

   Where:

   - `<YOUR-DB-HOSTNAME>` is your organization's assigned MongoDB Atlas instance hostname (including database name), as returned from `viam data database hostname` above.
   - `<YOUR-DB-USER>` is your organization's database user for that Atlas instance, as returned from `viam data database configure` above.

1. Once connected, you can run [MQL](https://www.mongodb.com/docs/manual/tutorial/query-documents/) or SQL to query captured data directly. For example:

   - The following MQL query searches the `sensorData` database and `readings` collection, and returns sensor readings from an ultrasonic sensor on a specific `robot_id` where the recorded `distance` measurement is greater than `.2` meters:

>      ```sql {class="line-numbers linkable-line-numbers"}

Reviewer: is the sql here just for markdown? should it be mql?

Author: Right on, I've just now [added](https://github.com/viamrobotics/docs/pull/2221) support for MQL syntax highlighting to our docs! 🎉

Reviewer: Awesome!!

---
title: "Configure Data Query"
linkTitle: "Configure Data Query"
description: "Configure data query to query tabular data with MQL or SQL"
weight: 35
type: "docs"
tags: ["data management", "cloud", "query", "sensor"]

> # SME: Aaron Casas

Reviewer: Can switch this to Devin Hilly

> # SME: Aaron Casas

Reviewer 2: ```suggestion
# SME: Devin Hilly
```

---

Configure data query to be able to directly query captured tabular data in the Viam cloud using [MQL](https://www.mongodb.com/docs/manual/tutorial/query-documents/) or SQL.
Synced data is stored in a MongoDB [Atlas Data Federation](https://www.mongodb.com/docs/atlas/data-federation/overview/) instance.

You can query against both the captured tabular data itself as well as its metadata, including robot ID, organization ID, and [tags](/manage/data/label/#image-tags).

Only tabular data, such as [sensor](/components/sensor/) readings, can be queried in this fashion.

## Requirements

Before you can configure data query, you must:

> - [Add the data management service](/services/data/configure-data-capture/#add-the-data-management-service)
> - [Configure data capture](/services/data/configure-data-capture/) for at least one component, such as a sensor.
>   Only components that capture tabular data support data query.
> - [Configure cloud sync](/services/data/configure-cloud-sync/), and sync data to the Viam app.
>   When you are able to [view your data in the Viam app](/manage/data/view/), you are ready to proceed.

Reviewer 2: ```suggestion
1. [Add the data management service](/services/data/configure-data-capture/#add-the-data-management-service)
1. [Configure data capture](/services/data/configure-data-capture/) for at least one component, such as a sensor.
   Only components that capture tabular data support data query.
1. [Configure cloud sync](/services/data/configure-cloud-sync/), and sync data to the Viam app.
   When you are able to [view your data in the Viam app](/manage/data/view/), you are ready to proceed.
```

## Configure data query

Once your smart machine has synced captured data to the Viam app, you can configure data query using the Viam CLI:

1. If you haven't already, [install the Viam CLI](/manage/cli/#install) and [authenticate](/manage/cli/#authenticate) to Viam.

1. Find your organization ID by running the following command, or from your organization's **Settings** page in [the Viam App](https://app.viam.com/):

   ```sh {class="line-numbers linkable-line-numbers"}
   viam organizations list
   ```

1. Configure a new database user for the Viam organization's MongoDB [Atlas Data Federation](https://www.mongodb.com/docs/atlas/data-federation/overview/) instance, which is where your smart machine's synced data is stored.
   Provide your organization's `org-id` from step 2, and a desired new password for your database user.

   ```sh {class="line-numbers linkable-line-numbers"}
   viam data database configure --org-id=<YOUR-ORGANIZATION-ID> --password=<NEW-DBUSER-PASSWORD>
   ```

   This command configures a new database user for your org for use with data query.
   If you have already created this user, this command updates the password for that user instead.

> 1. Determine the hostname for your organization's MongoDB Atlas Data Federation instance.
>    Provide your organization's `org-id` from step 2:

Reviewer 2: ```suggestion
1. Determine the hostname for your organization's MongoDB Atlas Data Federation instance by running the following command with the organization's `org-id` from step 2:
```

Use the `hostname` and `user` returned from the setup above to connect from your desired client to the MongoDB Atlas instance.

For example, to connect to your Viam organization's MongoDB Atlas instance and query data using the `mongosh` shell:

1. If you haven't already, [download the `mongosh` shell](https://www.mongodb.com/try/download/shell).
   See the [`mongosh` documentation](https://www.mongodb.com/docs/mongodb-shell/) for more information.

1. Run the following command to connect to the Viam organization's MongoDB Atlas instance from `mongosh`:

   ```sh {class="line-numbers linkable-line-numbers"}
   mongosh "mongodb+srv://<YOUR-DB-HOSTNAME>" --apiVersion 1 --username db-user-<YOUR-ORG-ID>
   ```

   Where:

   - `<YOUR-DB-HOSTNAME>` is your organization's assigned MongoDB Atlas instance hostname (including database name), as returned from `viam data database hostname` above.
   - `<YOUR-ORG-ID>` is your organization ID as determined above.

>      The full username you provide to the `--username` flag should therefore resemble `db-user-abcdef12-abcd-abcd-abcd-abcdef123456`.

Author: Q: correct? tech spec featured underescores.

Reviewer: This is correct

1. Once connected, you can run [MQL](https://www.mongodb.com/docs/manual/tutorial/query-documents/) or SQL to query captured data directly. For example:

>    - The following MQL query searches the `sensorData` database and `readings` collection, and returns sensor readings from an ultrasonic sensor on a specific `robot_id` where the recorded `distance` measurement is greater than `.2` meters:

Author: Q: Did I guess correctly that `.2` is unit: meters.

Reviewer: Correct!

     ```mongodb {class="line-numbers linkable-line-numbers"}
     AtlasDataFederation sensorData> db.readings.aggregate(
         [
             { $match: {
                 'robot_id': 'abcdef12-abcd-abcd-abcd-abcdef123456',

>                  'component_name': 'my-ultrasonic-sensor',

Author: Q: correct to infer component name here (i.e. with `my-`) vs component type (i.e. exactly `ultrasonic-sensor`)?

Reviewer: Yep you can name the component whatever you want.

                 'data.readings.distance': { $gt: .2 } } },
             { $count: 'numStanding' }
         ] )
     [ { numStanding: 215 } ]
     ```

   - The following SQL query uses the MongoDB [`$sql` aggregation pipeline stage](https://www.mongodb.com/docs/atlas/data-federation/query/sql/shell/connect/#aggregation-pipeline-stage-syntax) to perform the same query as the MQL above, but using SQL syntax:

     ```mongodb {class="line-numbers linkable-line-numbers"}
     AtlasDataFederation sensorData> db.aggregate(
         [

>              { $sql: {

Author: New `mongodb` syntax highlighting does not support `$sql` operator :(

---
title: "Configure Data Query"
linkTitle: "Configure Data Query"

> description: "Configure data query to query synced data directly in the cloud."

Reviewer: Perhaps `query tabular data with MQL or SQL`?

weight: 35
type: "docs"
tags: ["data management", "cloud", "query"]
# SME: Aaron Casas
---

Configure data query to be able to directly query captured data in the Viam cloud using the [MongoDB Query Language (MQL)](https://www.mongodb.com/docs/manual/tutorial/query-documents/).

## Requirements

Before you can configure [data query](../#data-query), you must [add the data management service](/services/data/configure-data-capture/#add-the-data-management-service) and [configure cloud sync](/services/data/configure-cloud-sync/).

## Configure data query

Once your smart machine has synced captured data to the Viam app, you can configure data query using the Viam CLI:

1. If you haven't already, [install the Viam CLI](/manage/cli/#install) and [authenticate](/manage/cli/#authenticate) to Viam.

> 1. Find your organization ID by running the following command, or from your organization's **Settings** page in [the Viam App](https://app.viam.com/):

Reviewer: [nit] are these numbers from here onwards supposed to increment, or does it render differently?

   ```sh {class="line-numbers linkable-line-numbers"}
   viam organizations list
   ```

1. Run the following command to determine the hostname for your organization's MongoDB [Atlas Data Federation](https://www.mongodb.com/docs/atlas/data-federation/overview/) instance, which is where your smart machine's synced data is stored.
   Provide your organizations `org-id` from step 2:

   ```sh {class="line-numbers linkable-line-numbers"}
   viam data database hostname --org-id=<YOUR-ORGANIZATION-ID>
   ```

   This command returns the `hostname` of your Viam organization's assigned MongoDB Atlas instance.

1. Then, configure a new database user for that database instance.
   Provide your organization's `org-id` from step 2, and a desired new password for your database user.

   ```sh {class="line-numbers linkable-line-numbers"}
   viam data database configure --org-id=<YOUR-ORGANIZATION-ID> --password=<NEW-DBUSER-PASSWORD>
   ```

   This command creates a new database `user` for your organization's MongoDB Atlas instance.

For more information, see the documentation for the [Viam CLI `database` command](/manage/cli/#data).

> ## Query data using MQL

Reviewer: I would probably mention that in this method of connecting directly to MongoDB, you can still use SQL queries by embedding them as $SQL in the Mongo query (example [here](https://www.mongodb.com/docs/atlas/data-federation/query/sql/shell/connect/#aggregation-pipeline-stage-syntax)).

Additionally, would soon (can be a fast-follow once it's in prod and not just in the deployment) add the main flow here that you can just input your SQL/MQL query into the app UI in order to query, and then have this be an alternative method if you want to connect directly to MongoDB.

For example, to connect to your Viam organization's MongoDB Atlas instance and query data using the `mongosh` shell:

1. If you haven't already, [download the `mongosh` shell](https://www.mongodb.com/try/download/shell).
   See the [`mongosh` documentation](https://www.mongodb.com/docs/mongodb-shell/) for more information.

1. Run the following command to connect to your Viam organization's MongoDB Atlas instance from `mongosh`:

   ```sh {class="line-numbers linkable-line-numbers"}
   mongosh "mongodb+srv://<YOUR-DB-HOSTNAME>" --apiVersion 1 --username <YOUR-DB-USER>
   ```

   Where:

   - `<YOUR-DB-HOSTNAME>` is your organization's assigned MongoDB Atlas instance, as returned from `viam data database hostname` above.
   - `<YOUR-DB-USER>` is your organization's database user for that Atlas instance, as returned from `viam data database configure` above.

1. Once connected, you can run MQL to query captured data directly.
   For example, the following MQL queries all synced data in the `sensors` database:

   ```sql {class="line-numbers linkable-line-numbers"}
   db.sensors.find( {} )
   ```

>    The following MQL queries synced data in the `sensors` database and returns `sensor_type: temp_sensor` readings greater than a certain `current_temperature`:

Reviewer: `sensorData` database and `readings` collection
